File: backend/app/schemas/assistant.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.services.assistant.handler import handle_assistant_request


class AssistantRequestIn(BaseModel):
    user_text: str = Field(default="", max_length=8000)
    selected_text: str | None = Field(default=None, max_length=32000)
    figure_id: str | None = Field(default=None, max_length=64)
    cursor_paragraph: str | None = Field(default=None, max_length=8000)
    explicit_intent: str | None = Field(default=None, max_length=64)
    chart_type: str | None = Field(default=None, max_length=32)
    sub_kind: str | None = Field(default=None, max_length=32)


class AssistantTextOut(BaseModel):
    type: str = "text"
    content: str
    intent: str | None = None


class AssistantFigureOut(BaseModel):
    type: str = "figure"
    figure_id: str
    file_url: str | None
    svg_url: str | None = None
    figure_number: str | None
    status: str
    caption: str | None
    figure_type: str
    updated_at: datetime | None = None
    intent: str | None = None

File: frontend/src/api/assistant.ts
import { client } from "@/api/client";

export type AssistantIntent =
  | "polish"
  | "rewrite"
  | "expand"
  | "condense"
  | "style_adjust"
  | "term_check"
  | "gen_flowchart"
  | "gen_chart"
  | "gen_figure"
  | "regen_figure";

export type AssistantRequest = {
  user_text?: string;
  selected_text?: string;
  figure_id?: string;
  cursor_paragraph?: string;
  explicit_intent?: AssistantIntent;
  chart_type?: string;
  sub_kind?: string;
};

export type AssistantTextResponse = {
  type: "text";
  content: string;
  intent?: string;
};

export type AssistantFigureResponse = {
  type: "figure";
  figure_id: string;
  file_url: string | null;
  svg_url?: string | null;
  figure_number: string | null;
  status: string;
  caption: string | null;
  figure_type: string;
  updated_at?: string | null;
  intent?: string;
};

export type AssistantConfirmResponse = {
  type: "confirm";
  message: string;
  intent?: string;
  confidence?: number;
  candidates?: string[];
};

export type AssistantResponse =
  | AssistantTextResponse
  | AssistantFigureResponse
  | AssistantConfirmResponse;

export async function callAssistant(
  bookId: string,
  chapterIndex: number,
  body: AssistantRequest,
): Promise<AssistantResponse> {
  const { data } = await client.post<AssistantResponse>(
    `/books/${bookId}/chapters/${chapterIndex}/assistant`,
    body,
    { timeout: 180000 },
  );
  return data;
}
